validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java

```diff
@@ -253,7 +262,36 @@ public void onAttestationEvent(BroadcastAttestationEvent event) throws IllegalAr
       // At the start of each epoch or at genesis, update attestation assignments
       // for all validators
       if (isGenesisOrEpochStart(slot)) {
-        updateAttestationAssignments(headState);
+        List<Integer> committeeIndices = updateAttestationAssignments(headState);
+        List<Integer> committeeIndicesToRegister = new ArrayList<>();
+
+        committeeIndices.forEach(
+            index -> {
+              if (!committeeIndexTTL.containsKey(index)) {
+                committeeIndicesToRegister.add(index);
+              }
+              committeeIndexTTL.put(index, COMMITTEE_INDEX_SUBSCRIPTION_LENGTH);
+            });
+        if (!committeeIndicesToRegister.isEmpty()) {
+          this.eventBus.post(new CommitteeAssignmentEvent(committeeIndicesToRegister));
+        }
+
+        List<Integer> committeeIndicesToDeregister = new ArrayList<>();
+        for (Map.Entry<Integer, Integer> entry : committeeIndexTTL.entrySet()) {
+          int ttl = entry.getValue();
+          int index = entry.getKey();
+          ttl--;
+          if (ttl == 0) {
+            committeeIndicesToDeregister.add(index);
+            committeeIndexTTL.remove(index);
+          } else {
+            entry.setValue(ttl);
+          }
```

Comment:
This would be a lot simpler if instead of storing a map of TTLs we just stored the slot at which we should deregister the index.

So when registering:
`put(index, slot + SLOTS_PER_EPOCH * COMMITTEE_INDEX_SUBSCRIPTION_LENGTH)`

and then you remove and deregister if the current slot > slotInMap.

Reply:
I'd also suggest that this code to send subscribe and unsubscribe events should be split out into a separate class and then you can unit test it much more easily.

Something like:
```
public class CommitteeAssignmentManager {
  private final Map<Integer, Integer> committeeIndexToDeregisterSlot = new HashMap<>(); 
  private final EventBus eventBus; // set in constructor

  public void updateCommitteeAssignments(UnsignedLong slot, List<Integer> committeeIndices) {...}
}
```
The unit tests can then listen to the event bus to check that the right `CommitteeAssignmentEvent` and `CommitteeDismissalEvent` events are fired.

Reply:
Great ideas. I'm integrating them into the PR right now.

Reply:
Done.